hu.elte.refjava.lang/src/hu/elte/refjava/api/ClassRefactoring.xtend

```diff
@@ -0,0 +1,255 @@
+
+	protected new(String matchingPatternString, String replacementPatternString) {
+		matcher = new PatternMatcher(PatternParser.parse(matchingPatternString))
+		builder = new ASTBuilder(PatternParser.parse(replacementPatternString))
+		this.matchingString = matchingPatternString
+		this.replacementString = replacementPatternString
+	}
+	
+	override init(List<? extends ASTNode> target, IDocument document, List<TypeDeclaration> allTypeDeclInWorkspace) {
+		this.target = target
+		this.document = document
+		this.typeDeclList = allTypeDeclInWorkspace
+	}
+	
+	def setTarget(List<?extends ASTNode> newTarget) {
+		this.target = newTarget
+	}
+	
+	override apply() {
+		setMetaVariables()
+		return if(!safeTargetCheck) {
+			Status.TARGET_MATCH_FAILED
+		} else if (!safeMatch) {
+			Status.MATCH_FAILED
+		} else if (!safeCheck) {
+			Status.CHECK_FAILED
+		} else if (!safeBuild) {
+			Status.BUILD_FAILED
+		} else if (!safeReplace) {
+			Status.REPLACEMENT_FAILED
+		} else {
+			Status.SUCCESS
+		}
+	}
+	
+	def private safeMatch() {
+		try {
+			if (!matcher.match(target, nameBindings, typeBindings, parameterBindings, matchingTypeReferenceString)) {
+				return false
+			}
+			target = matcher.modifiedTarget.toList
+			bindings.putAll(matcher.bindings)
+			
+			return true
+		} catch (Exception e) {
+			println(e)
+			return false
+		}
+	}
+	
+	def protected void setMetaVariables() {
+		//empty
+	}
+	
+	def protected safeTargetCheck() {
+		return true
+	}
+	
+	def protected targetCheck(String targetPatternString) {
+		try {
+			this.targetString = targetPatternString
+			if(!matcher.match(PatternParser.parse(targetPatternString), target, targetTypeReferenceString)) {
+				return false
+			}
+			bindings.putAll(matcher.bindings)
+			return true
+		} catch (Exception e) {
+			println(e)
+			return false
+		}	
+	}
+
+	def private safeCheck() {
+		try {
+			return check()
+		} catch (Exception e) {
+			println(e)
+			return false
+		}
+	}
+
+	def protected check() {
+		//TODO
+		if(replacementString != "-") {
+			//return Check.isUniqueMethodIn(target, Check.enclosingClass(target))
+			
+			return true
+			
+			
+		} else {
+			if(target.head instanceof MethodDeclaration) {
+				return Check.hasSuperClass(Check.enclosingClass(target))
+					&& if (Check.isPrivate(target)) Check.references(target, Check.enclosingClass(target)).empty else true
+					&& Check.accessedFieldsOfEnclosingClass(target, Check.enclosingClass(target)).empty
+					&& Check.accessedMethodsOfEnclosingClass(target, Check.enclosingClass(target)).empty
+				//	&& Check.isUniqueMethodIn(target, Check.superClass(Check.enclosingClass(target)))
+					
+				
+			} else if (target.head instanceof FieldDeclaration) {
+				return Check.hasSuperClass(Check.enclosingClass(target))
+					&& if (Check.isPrivate(target)) Check.references(target, Check.enclosingClass(target)).empty else true
+				//	&& Check.isUniqueFieldIn(target, Check.superClass(Check.enclosingClass(target)))
+			}
```

Comment:
The preconditions of the _introduce method into the enclosing class_ refactoring seem to be missing, but I assume that's covered under the starting `//TODO`. Nevertheless, it might be desirable to organize the 3 kinds of the class scheme into 3 subclasses of `ClassRefactoring`. What do you think?